deps/media-playback/media-playback/media.c

```diff
@@ -258,125 +258,183 @@ static inline bool mp_media_can_play_frame(mp_media_t *m,
 static void mp_media_next_audio(mp_media_t *m)
 {
 	struct mp_decode *d = &m->a;
-	struct obs_source_audio audio = {0};
 	AVFrame *f = d->frame;
 
-	if (!mp_media_can_play_frame(m, d))
-		return;
+	if (m->audio.index_eof > 0 && m->audio.index == m->audio.index_eof) {
+		m->audio.index = 0;
+		m->next_wait = 0;
+	}
 
-	d->frame_ready = false;
-	if (!m->a_cb)
-		return;
+	if (m->audio.index_eof < 0) {
 
-	for (size_t i = 0; i < MAX_AV_PLANES; i++)
-		audio.data[i] = f->data[i];
+		if (!mp_media_can_play_frame(m, d))
+			return;
 
-	audio.samples_per_sec = f->sample_rate * m->speed / 100;
-	audio.speakers = convert_speaker_layout(f->channels);
-	audio.format = convert_sample_format(f->format);
-	audio.frames = f->nb_samples;
-	audio.timestamp = m->base_ts + d->frame_pts - m->start_ts +
-		m->play_sys_ts - base_sys_ts;
+		d->frame_ready = false;
+		if (!m->a_cb)
+			return;
 
-	if (audio.format == AUDIO_FORMAT_UNKNOWN)
-		return;
+		struct obs_source_audio *audio = malloc(sizeof(struct obs_source_audio));;
 
-	m->a_cb(m->opaque, &audio);
+		for (size_t i = 0; i < MAX_AV_PLANES; i++) {
+			audio->data[i] = malloc(f->linesize[0] / f->channels);
+			if (f->data[i]) {
+				memcpy(audio->data[i], f->data[i], f->linesize[0] / f->channels);
+			}
+		}
+
+		audio->samples_per_sec = f->sample_rate * m->speed / 100;
+		audio->speakers = convert_speaker_layout(f->channels);
+		audio->format = convert_sample_format(f->format);
+		audio->frames = f->nb_samples;
+		audio->timestamp = m->base_ts + d->frame_pts - m->start_ts +
+			m->play_sys_ts - base_sys_ts;
+
+		if (audio->format == AUDIO_FORMAT_UNKNOWN)
+			return;
```

Comment:
"audio" and "audio->data[]" can be leaked here?

Reply:
No they should be correctly released in the `clear_cache` method

Reply:
to be removed in clear_cache they have to be saved "m->audio". 
But if function exit on "if (audio->format == AUDIO_FORMAT_UNKNOWN) return ;" then "audio" doesn't have a chance to be saved to it.

Reply:
should be fixed